docs/tutorials/tutorial_2.rst

```diff
@@ -0,0 +1,537 @@
+            U.push_back(i);
+            V.push_back(rnd.nextInt(0, i - 1));
+        }
+        renumber(N, U, V);
+    }
+
+    void randomGraph(int N, int M, vector<int> &U, vector<int> &V) {
+        randomTree(N, U, V);
+        while (U.size() < M) {
+            int u = rnd.nextInt(0, N - 2);
+            int v = rnd.nextInt(u + 1, N - 1);
+            U.push_back(u);
+            V.push_back(v);
+        }
+    }
+
+Now, let us move on to the creating test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the venn diagram of the subtasks.
+
+The properties of the subtasks of the above problem are :
+
+1. All testcases inside subtask 1 and subtask 2 are also inside subtask 3, and subtask 3 have more testcases that is neither in subtask 1 nor subtask 2.
+
+2. Some testcases are inside both subtask 1 and subtask 2, and there are also testcases that is in subtask 1 but not in subtask 2, and there are also testcases that is in subtask 2 but not in subtask 2.
+
+Therefore, the Venn diagram looks like this
+
+.. image:: tutorial_2-venn_diagram.png
+
+We need to define a test group for each of the region in the Venn diagram. Therefore, we will have four testgroups. Let us number it from 1 to 4 in the same order as the diagram above. Therefore, each testgroups will have the following constraints in addition to the global constraints:
+
+1. M ≤ 20 and all W are equal
+2. M ≤ 20 and not all W are equal
+3. M > 20 and all W are equal
+4. M > 20 and not all W are equal
+
+Therefore, the test cases for the first test group looks something like this
+
+.. sourcecode:: cpp
+
+    void TestGroup1() {
+        Subtasks({1, 2, 3});
+
+        CASE(N = 2, M = 1, U = {0}, V = {1}, W = {1});
+        CASE(N = 21, M = 20, randomTree(N, U, V), W.assign(M, 1000));
+        CASE(N = 20, M = 20, randomGraph(N, M, U, V), W.assign(M, 1000));
+
+        for (int i = 0; i < 5; ++i) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 W.assign(M, rnd.nextInt(1, 1000)));
+        }
+    }
+
+while the second test group looks something like this
+
+.. sourcecode:: cpp
+
+    void TestGroup2() {
+        Subtasks({1, 3});
+        
+        CASE(N = 4, M = 4,
+             U = {0, 1, 2, 0},
+             V = {1, 2, 0, 3},
+             W = {1, 1, 1, 2});
+
+        CASE(N = 2, M = 2, U = {0, 1}, V = {1, 0}, W = {1, 2});
+        CASE(N = 21, M = 20, randomTree(N, U, V), randomWeight(M, W));
+
+        for (int i = 0; i < 5; ++i) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 randomWeight(M, W));
+        }
+    }
+
+and so on. Note that in the second testgroup, we manually create a small test case where greedily choosing the first N - 1 edges with smallest weight will create a cycle.
```

Comment:
Put this directly as a comment in the source code instead. Also, it's better to put comment for every (group of) test case for more clarity, in general